src/data/dfinsupp/lex.lean

```diff
@@ -0,0 +1,165 @@
+{ irrefl := to_lex.surjective.forall.2 $ λ a, @irrefl _ _ i.to_is_irrefl a,
+  trans := to_lex.surjective.forall₃.2 $ λ a b c, @trans _ _ i.to_is_trans a b c }
+
+variables [linear_order ι]
+
+/-- The partial order on `dfinsupp`s obtained by the lexicographic ordering.
+See `dfinsupp.lex.linear_order` for a proof that this partial order is in fact linear. -/
+instance lex.partial_order [Π i, partial_order (α i)] : partial_order (lex (Π₀ i, α i)) :=
+partial_order.lift (λ x, to_lex ⇑(of_lex x)) dfinsupp.coe_fn_injective
+
+section linear_order
+
+variable [Π i, linear_order (α i)]
+
+/-- Auxiliary helper to case split computably. There is no need for this to be public, as it
+can be written with `or.by_cases` on `lt_trichotomy` once the instances below are constructed. -/
+private def lt_trichotomy_rec {P : lex (Π₀ i, α i) → lex (Π₀ i, α i) → Sort*}
+  (h_lt : Π {f g}, to_lex f < to_lex g → P (to_lex f) (to_lex g))
+  (h_eq : Π {f g}, to_lex f = to_lex g → P (to_lex f) (to_lex g))
+  (h_gt : Π {f g}, to_lex g < to_lex f → P (to_lex f) (to_lex g)) :
+  Π f g, P f g :=
+lex.rec $ λ f, lex.rec $ λ g,
+  match _, rfl : ∀ y, (f.ne_locus g).min = y → _ with
+  | ⊤, h := h_eq (ne_locus_eq_empty.mp $ finset.min_eq_top.mp h)
+  | (wit : ι), h := (mem_ne_locus.mp $ finset.mem_of_min h).lt_or_lt.by_cases
+      (λ hwit, h_lt ⟨wit, λ j hj, not_mem_ne_locus.mp (finset.not_mem_of_lt_min hj h), hwit⟩)
+      (λ hwit, h_gt ⟨wit, λ j hj, not_mem_ne_locus.mp
+        (finset.not_mem_of_lt_min hj $ by rwa ne_locus_comm), hwit⟩)
+  end
+
+@[irreducible] instance lex.decidable_le : @decidable_rel (lex (Π₀ i, α i)) (≤) :=
+lt_trichotomy_rec
+  (λ f g h, is_true $ or.inr h)
+  (λ f g h, is_true $ or.inl $ congr_arg _ h)
+  (λ f g h, is_false $ λ h', (lt_irrefl _ (h.trans_le h')).elim)
+
+@[irreducible] instance lex.decidable_lt : @decidable_rel (lex (Π₀ i, α i)) (<) :=
+lt_trichotomy_rec
+  (λ f g h, is_true h)
+  (λ f g h, is_false h.not_lt)
+  (λ f g h, is_false h.asymm)
+
+/--  The linear order on `dfinsupp`s obtained by the lexicographic ordering. -/
+instance lex.linear_order : linear_order (lex (Π₀ i, α i)) :=
+{ le_total := lt_trichotomy_rec
+    (λ f g h, or.inl h.le)
+    (λ f g h, or.inl h.le)
+    (λ f g h, or.inr h.le),
+  decidable_lt := by apply_instance,
+  decidable_le := by apply_instance,
+  decidable_eq := by apply_instance,
+  ..lex.partial_order }
+
+end linear_order
+
+variable [Π i, partial_order (α i)]
+
+lemma lex.le_of_forall_le {a b : lex (Π₀ i, α i)} (h : ∀ i, of_lex a i ≤ of_lex b i) : a ≤ b :=
+le_of_lt_or_eq $ or_iff_not_imp_right.2 $ λ hne, by classical; exact
+  ⟨finset.min' _ (nonempty_ne_locus_iff.2 hne),
+    λ j hj, not_mem_ne_locus.1 (λ h, (finset.min'_le _ _ h).not_lt hj),
+    (h _).lt_of_ne (mem_ne_locus.1 $ finset.min'_mem _ _)⟩
+
+lemma lex.le_of_of_lex_le {a b : lex (Π₀ i, α i)} (h : of_lex a ≤ of_lex b) : a ≤ b :=
+lex.le_of_forall_le h
+
+lemma to_lex_monotone : monotone (@to_lex (Π₀ i, α i)) :=
+λ _ _, lex.le_of_forall_le
```

Comment:
Out of these four, the only one I would expect is `to_lex_monotone` and I would add its counterpart `to_lex_strict_mono`. The other ones have confusing names and if I were to discover the API for the first time I would wonder what their use is given that we already have `to_lex_monotone` and `to_lex_strict_mono`.

`lex_lt_of_lt_of_preorder` and `lex_lt_of_lt` are similarly unsettling because they mix bundled relations with unbundled ones.

Reply:
Git blame tells me that all three were introduced by @urkud for `pi.lex`, with [the first one](https://github.com/leanprover-community/mathlib/commit/6a6b0a565bc853c356f5947fd9ac49e167a67e2c#diff-edbaf27dd2f04d3c59f00189352e2e4e6ce429f38db7a6c86c8d9a89d211d6d9R72-R74) in #9129 and [the other two](https://github.com/leanprover-community/mathlib/commit/d136cd555dab1b0781eb7280e976d42d91cd9e92#diff-7afc24d61cb458dafb7dc06846ecaa2dbbc235bceb031d3b2fee998e806f21abR89-R99) in #14389, and you've also touched the first one in #14164. @adomani was probably just copying from `pi.lex` like what I do here. I'm happy to have the first two removed because all are defeq.

Reply:
Just to confirm: I did simply copy a bunch of lemmas and fix as instructed by Lean!

Reply:
> lex_lt_of_lt_of_preorder and lex_lt_of_lt are similarly unsettling because they mix bundled relations with unbundled ones.

It's the only way to state them succinctly; I'd like to use unbundled relation for the order on `α i` as well, but although `a ≤ a'` is just `∀ i, a i ≤ a' i` (product order), `a < a'` is defined to be `a ≤ a' ∧ a' ≰ a`; we don't have a definition for this operation on relations, and I think writing `(∀ i, r (a i) (a' i)) ∧ (∃ i, ¬ r (a' i) (a i))` is too long, so I rely on the typeclass system to generate the (<) for me, and that requires the use of `preorder` for the relations on `α i`.

And really these are just 3 x 2 technical lemmas used to transfer well-foundedness of lex order to product order (in fact only the first of the dfinsupp lemmas is actually used, I believe).

Reply:
Hi @YaelDillies I've removed the 3x2 lemmas lex.le_of_forall_le/le_of_of_lex_le from pi/finsupp/dfinsupp and CI now passed. I also renamed the dfinsupp.ne_locus lemmas following your PR #17036. So the only thing not yet resolved is about `lex_lt_of_lt(_of_preorder)`; do you have a better suggestion, or should I leave them as is?